src/openms/source/ANALYSIS/OPENSWATH/DIAHelper.cpp

```diff
@@ -58,143 +58,290 @@ namespace OpenMS::DIAHelpers
       {
         left -= mz_extract_window / 2.0;
         right += mz_extract_window / 2.0;
+
+        // If the left value is now < 0, this is invalid correct it to be 0
+        if (left < 0)
+        {
+          left = 0;
+        }
       }
     }
 
+    // Helper for integrate window returns the sum of all intensities, sum of all ion mobilities and sum of all mz
+    // no expensive division calls
+    // assumes mz, im and intensity should already be initiated.
+    void _integrateWindowHelper(const OpenSwath::SpectrumPtr& spectrum,
+                double mz_start,
+                double mz_end,
+                double & mz,
+                double & im,
+                double & intensity,
+                double drift_start,
+                double drift_end,
+                bool centroided)
+    {
+      // TODO are preconditions working?
+      OPENMS_PRECONDITION(spectrum != nullptr, "Spectrum cannot be nullptr");
+      OPENMS_PRECONDITION(spectrum->getMZArray() != nullptr, "Cannot integrate if no m/z is available.");
+      OPENMS_PRECONDITION(spectrum->getMZArray()->empty(), "Cannot integrate if no m/z is available.");
+      OPENMS_PRECONDITION(std::adjacent_find(spectrum->getMZArray()->data.begin(),
+        spectrum->getMZArray()->data.end(), std::greater<double>()) == spectrum->getMZArray()->data.end(),
+        "Precondition violated: m/z vector needs to be sorted!" );
+      OPENMS_PRECONDITION(spectrum->getMZArray()->data.size() == spectrum->getIntensityArray()->data.size(), "MZ and Intensity array need to have the same length.");
+      OPENMS_PRECONDITION(!centroided,  throw Exception::NotImplemented(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION));
+
+      if (spectrum->getMZArray()->data.empty())
+      {
+        std::cerr << "Warning: Cannot integrate if spectrum is empty" << std::endl;
+        return;
+      }
+
+      // drift time checks
+      if ( drift_start >= 0 ) // integrate across im as well
+      {
+        // additional checks only relevant if ion mobility is present
+
+        //TODO this preconditions do not seem to be working? however the if statement below works
+        OPENMS_PRECONDITION(spectrum->getDriftTimeArray() != nullptr, "Cannot integrate with drift time if no drift time is available.");
+        OPENMS_PRECONDITION(spectrum->getMZArray()->data.size() == spectrum->getDriftTimeArray()->data.size(), "MZ and Drift Time array need to have the same length.");
+
+        if (spectrum->getDriftTimeArray() == nullptr)
+        {
+          std::cerr << "Warning: Cannot integrate with drift time if no drift time is available.\n";
+          return;
```

Comment:
is this a good way to deal with the error? what happens now to the function calling? should we instead throw an exception?